libtenzir/builtins/operators/timeshift.cpp

```diff
@@ -0,0 +1,160 @@
+      if (not resolved_field->second) {
+        co_yield std::move(slice);
+        continue;
+      }
+      auto transform_fn = [&](struct record_type::field field,
+                              std::shared_ptr<arrow::Array> array) noexcept
+        -> indexed_transformation::result_type {
+        TENZIR_ASSERT_CHEAP(caf::holds_alternative<time_type>(field.type));
+        auto builder = series_builder{};
+        for (auto element : values(
+               time_type{},
+               static_cast<const type_to_arrow_array_t<time_type>&>(*array))) {
+          if (not element) {
+            builder.null();
+            continue;
+          }
+          if (not first_time) [[unlikely]] {
+            first_time = *element;
+            if (not start) {
+              start = *element;
+            }
+          }
+          builder.data(*start
+                       + std::chrono::duration_cast<duration>(
+                         (*element - *first_time) / speed_));
+        }
+        auto arrays = builder.finish();
+        TENZIR_ASSERT_CHEAP(arrays.size() == 1);
+        TENZIR_ASSERT_CHEAP(caf::holds_alternative<time_type>(arrays[0].type));
+        return {{field, std::move(arrays[0].array)}};
+      };
+      auto transformations = std::vector<indexed_transformation>{
+        {*resolved_field->second, transform_fn},
+      };
+      co_yield transform_columns(slice, transformations);
+    }
+  }
+
+  auto optimize(expression const& filter, event_order order) const
+    -> optimize_result override {
+    if (speed_ == 1.0 and not start_) {
+      // If this operator is a no-op we can just remove it during optimization.
+      return optimize_result{filter, order, nullptr};
+    }
+    return do_not_optimize(*this);
+  }
+
+  friend auto inspect(auto& f, timeshift_operator& x) -> bool {
+    return f.object(x)
+      .pretty_name("tenzir.plugins.timeshift.timeshift_operator")
+      .fields(f.field("field", x.field_), f.field("speed", x.speed_),
+              f.field("start", x.start_));
+  }
+
+private:
+  std::string field_ = {};
+  double speed_ = 1.0;
+  std::optional<time> start_ = {};
+};
+
+class plugin final : public virtual operator_plugin<timeshift_operator> {
+public:
+  auto signature() const -> operator_signature override {
+    return {.transformation = true};
+  }
+
+  auto parse_operator(parser_interface& p) const -> operator_ptr override {
+    auto speed = std::optional<double>{};
+    auto start = std::optional<time>{};
+    auto field = std::string{};
+    auto parser = argument_parser{"timeshift", "https://docs.tenzir.com/next/"
+                                               "operators/timeshift"};
+    parser.add("--speed", speed, "<factor>");
+    parser.add("--start", start, "<time>");
+    parser.add(field, "<field>");
+    parser.parse(p);
+    if (speed and *speed <= 0.0) {
+      diagnostic::error("`--speed` must be greater than 0")
+        .note("from `{}`", name())
+        .throw_();
+    }
```

Comment:
We can error here if `not start and not speed` to make the user aware that it's a no-op. I don't have a strong opinion on that.

Reply:
I'd say leave it as is now.